### PR TITLE
Use streams to avoid multiple queries during batch job id chunking.

### PR DESCRIPTION
**Pull request by michaelabuckley**
We make repeated queries that simulate paging when walking a query url for a batch job.  This change replaces that with a streaming api so we execute a single query and reduce load on the database.

### CODE REVIEW

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/pid/AutoClosingStreamTemplate.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Template for wrapping access to stream supplier in a try-with-resources block.
+ */
+class AutoClosingStreamTemplate<T> implements StreamTemplate<T> {
+	private final Supplier<Stream<T>> theStreamQuery;
+
+	AutoClosingStreamTemplate(Supplier<Stream<T>> theTheStreamQuery) {
```
Comment (lukedegruchy) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/pid/AutoClosingStreamTemplate.java:15]: Nitpick:  theStreamQuery without an extra "the"

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/pid/TransactionWrappingStreamTemplate.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * Wrap a StreamTemplate with transaction advice.
+ * We can't cary open ResultSets past a transaction boundary.
+ * This wraps a Stream producer with tx advice so the connection is still open.
+ */
+class TransactionWrappingStreamTemplate<T> implements StreamTemplate<T> {
+	@Nonnull
+	final TransactionOperations myTransaction;
+
+	@Nonnull
+	final StreamTemplate<T> theWrappedStreamTemplate;
```
Comment (lukedegruchy) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/pid/TransactionWrappingStreamTemplate.java:20]: Nitpick:  myWrappedStreamTemplate

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/pid/TransactionWrappingStreamTemplate.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+	TransactionWrappingStreamTemplate(
+			@Nonnull TransactionOperations theTransaction, @Nonnull StreamTemplate<T> theTheWrappedStreamTemplate) {
```
Comment (lukedegruchy) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/pid/TransactionWrappingStreamTemplate.java:23]: Nitpick:  theWrappedStreamTemplate with an extra "the"

File: hapi-fhir-storage-batch2/src/test/java/ca/uhn/fhir/batch2/jobs/step/LoadIdsStepTest.java [hunk 4]
```diff
@@ -35,6 +35,7 @@
 
 @ExtendWith(MockitoExtension.class)
 public class LoadIdsStepTest {
+	// wipmb review
```
Comment (lukedegruchy) [inline hapi-fhir-storage-batch2/src/test/java/ca/uhn/fhir/batch2/jobs/step/LoadIdsStepTest.java:38]: Do you still need this?

File: hapi-fhir-storage-batch2/src/test/java/ca/uhn/fhir/batch2/jobs/step/ResourceIdListStepTest.java [hunk 5]
```diff
@@ -34,6 +36,7 @@
 
 @ExtendWith(MockitoExtension.class)
 class ResourceIdListStepTest {
+	// wipmb review
```
Comment (lukedegruchy) [inline hapi-fhir-storage-batch2/src/test/java/ca/uhn/fhir/batch2/jobs/step/ResourceIdListStepTest.java:39]: Do you still need this?

File: hapi-fhir-storage-batch2/src/test/java/ca/uhn/fhir/batch2/jobs/step/ResourceIdListStepTest.java [hunk 6]
```diff
@@ -55,31 +58,29 @@ void beforeEach() {
 		myResourceIdListStep = new ResourceIdListStep<>(myIdChunkProducer);
 	}
 
+	// wipmb can we delete this?
```
Comment (lukedegruchy) [inline hapi-fhir-storage-batch2/src/test/java/ca/uhn/fhir/batch2/jobs/step/ResourceIdListStepTest.java:61]: Are you planning to address these comments in this PR?

File: hapi-fhir-storage-mdm/src/main/java/ca/uhn/fhir/mdm/batch2/MdmIdChunkProducer.java [hunk 7]
```diff
@@ -55,4 +59,13 @@ public IResourcePidList fetchResourceIdsPage(
 		return myGoldenResourceSearchSvc.fetchGoldenResourceIdsPage(
 				theNextStart, theEnd, thePageSize, theRequestPartitionId, resourceType);
 	}
+
+	// wipmb convert this to native stream. Maybe just call batch2Daosvcimpl with Resource?_tag=blah,blah
```
Comment (lukedegruchy) [inline hapi-fhir-storage-mdm/src/main/java/ca/uhn/fhir/mdm/batch2/MdmIdChunkProducer.java:63]: Question: Is the plan to make equivalent changes in cdr in an associated MR?

File: hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/GoldenResourceSearchSvcImpl.java [hunk 8]
```diff
@@ -56,24 +59,31 @@ public class GoldenResourceSearchSvcImpl implements IGoldenResourceSearchSvc {
 	@Autowired
 	private FhirContext myFhirContext;
 
+	@Autowired
+	private IHapiTransactionService myTransactionService;
+
 	@Override
 	@Transactional
-	public IResourcePidList fetchGoldenResourceIdsPage(
+	public IResourcePidStream fetchGoldenResourceIdStream(
```
Comment (lukedegruchy) [inline hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/GoldenResourceSearchSvcImpl.java:67]: Question:  What's the use case for applying this new framework to MDM as well?

**Review by lukedegruchy — APPROVED** [review]

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/pid/AutoClosingStreamTemplate.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Template for wrapping access to stream supplier in a try-with-resources block.
+ */
```
Comment (lukedegruchy) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/pid/AutoClosingStreamTemplate.java:11]: Please add copyright headers to all new classes.

File: hapi-fhir-storage-batch2/src/test/java/ca/uhn/fhir/batch2/jobs/step/ResourceIdListStepTest.java (hunk 5 above)
Comment (michaelabuckley) [reply hapi-fhir-storage-batch2/src/test/java/ca/uhn/fhir/batch2/jobs/step/ResourceIdListStepTest.java]: Removed

File: hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/GoldenResourceSearchSvcImpl.java (hunk 8 above)
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-mdm/src/main/java/ca/uhn/fhir/jpa/mdm/svc/GoldenResourceSearchSvcImpl.java:67]: This is for the mdm-clear batch job.  It has the same problem - repeated queries with a sort.  Streaming one long query is faster and easier on the db.

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/pid/AutoClosingStreamTemplate.java (hunk 9 above)
Comment (michaelabuckley) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/pid/AutoClosingStreamTemplate.java:30]: done.

**Pull request merged**